scipy/stats/qmc.py

```diff
@@ -0,0 +1,1368 @@
+    dim : int
+        Dimension of the parameter space.
+    seed : {None, int, `~np.random.RandomState`, `~np.random.Generator`}, optional
+        This parameter defines the object to use for drawing random variates.
+        If `seed` is `None` the `~np.random.RandomState` singleton is used.
+        If `seed` is an int, a new ``RandomState`` instance is used, seeded
+        with seed.
+        If `seed` is already a ``RandomState`` or ``Generator`` instance,
+        then that object is used.
+        Default is None.
+
+    Notes
+    -----
+
+    By convention samples are distributed over the half-open interval ``[0, 1)``.
+    Instances of the class can access the attributes: ``dim`` for the dimension;
+    and ``rng`` for the random number generator (used for the ``seed``).
+
+    **Subclassing**
+
+    When subclassing `QMCEngine` to create a new sampler,  ``__init__`` and
+    ``random`` has to be redefined.
+
+    * ``__init__(dim, seed=None)``: at least fix the dimension. If the sampler
+      does not take advantage of a ``seed`` (deterministic methods like Halton),
+      this parameter can be omitted.
+    * ``random(n_samples)``: draw ``n_samples`` from the engine.
+
+    Optionally, 2 other methods can be overwritten by subclasses:
+
+    * ``reset``: Reset the engine to it's original state.
+    * ``fast_forward``: It should be used as a way to fast-forward a sequence
+      to a further state. If the sequence is deterministic (like Halton
+      sequence), then ``fast_forward(n)`` is skipping the ``n`` first draw.
+
+    Examples
+    --------
+    To create a random sampler based on ``np.random.random``, we would do the
+    following:
+
+    >>> from scipy.stats import qmc
+    >>> class RandomEngine(qmc.QMCEngine):
+    ...     def __init__(self, dim, seed):
+    ...         super().__init__(dim=dim, seed=seed)
+    ...
+    ...
+    ...     def random(self, n_samples=1):
+    ...         return self.rng.random((n_samples, self.dim))
+    ...
+
+    We subclass `QMCEngine` by defining the sampling strategy we want to use.
+    And we can create an instance to sample from.
+
+    >>> engine = RandomEngine(2, seed=12345)
+    >>> engine.random(5)
+    array([[0.92961609, 0.31637555],
+           [0.18391881, 0.20456028],
+           [0.56772503, 0.5955447 ],
+           [0.96451452, 0.6531771 ],
+           [0.74890664, 0.65356987]])
+
+    """
+
+    @abstractmethod
+    def __init__(self, dim, seed=None):
+        self.dim = dim
+        self.rng = check_random_state(seed)
+        self.num_generated = 0
+
+    @abstractmethod
+    def random(self, n_samples=1):
+        """Draw `n_samples` in the half-open interval ``[0, 1)``.
+
+        Parameters
+        ----------
+        n_samples : int
+            Number of samples to generate in the parameter space.
+
+        Returns
+        -------
+        sample : array_like (n_samples, dim)
+            QMC sample.
+        """
+        #self.num_generated += n_samples
```

Comment:
I agree. The only reason I did this was to achieve better coverage. The thing is that this is never called as being an abstract method. If you prefer I add a pass. And maybe you have a way to explicitly skip these for coverage??

Reply:
I see, as long as nobody else is worried you may be able to just leave it that way, I had just never seen it before so was curious.

There's no way to have an issue where it gets called through `super()` or similar when subclassing?

Reply:
I just tested and you can do this:

```python
class Toto(qmc.QMCEngine):
    def __init__(self, dim):
        super().__init__(dim=dim)

    def random(self):
        super().random()
        
a = Toto(2)
```

But currently I would not see the point of doing this.

Otherwise, I think it's fine considering they are doing this in scikit-learn (here for instance): https://github.com/scikit-learn/scikit-learn/blob/5756a9934b2859b3525b76164ba3ec95a30a03f1/sklearn/decomposition/_base.py

As for the official Python doc, they use an ellipsis. And I found some options here to exclude from coverage if we need to: https://stackoverflow.com/q/9202723/6522112